lxd/auth/driver_openfga.go

```diff
@@ -0,0 +1,1079 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/openfga/go-sdk"
+	"github.com/openfga/go-sdk/client"
+	"github.com/openfga/go-sdk/credentials"
+
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+	"github.com/canonical/lxd/shared/logger"
+)
+
+func WriteOpenFGAAuthorizationModel(ctx context.Context, apiURL string, apiToken string, storeID string) (string, error) {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	u, err := url.Parse(apiURL)
+	if err != nil {
+		return "", fmt.Errorf("Failed parsing URL: %w", err)
+	}
+
+	conf := client.ClientConfiguration{
+		ApiScheme: u.Scheme,
+		ApiHost:   u.Host,
+		StoreId:   storeID,
+		Credentials: &credentials.Credentials{
+			Method: credentials.CredentialsMethodApiToken,
+			Config: &credentials.Config{
+				ApiToken: apiToken,
+			},
+		},
+	}
+
+	fgaClient, err := client.NewSdkClient(&conf)
+	if err != nil {
+		return "", fmt.Errorf("Failed to create OpenFGA client: %w", err)
+	}
+
+	var builtinAuthorizationModel client.ClientWriteAuthorizationModelRequest
+	err = json.Unmarshal([]byte(authModel), &builtinAuthorizationModel)
+	if err != nil {
+		return "", fmt.Errorf("Failed to unmarshal built in authorization model: %w", err)
+	}
+
+	data, err := fgaClient.WriteAuthorizationModel(ctx).Body(builtinAuthorizationModel).Execute()
+	if err != nil {
+		return "", err
+	}
+
+	return data.GetAuthorizationModelId(), nil
+}
+
+type fga struct {
+	commonAuthorizer
+
+	apiURL      string
+	apiToken    string
+	storeID     string
+	authModelID string
+
+	client *client.OpenFgaClient
+}
+
+func (f *fga) configure(opts Opts) error {
```

Comment:
The type assertion `val.(string)` could potentially panic if the underlying type of val is not a string. It might be safer to use a type assertion with a check, like:
```go
strVal, ok := val.(string)
if !ok {
    return fmt.Errorf("Expected a string for configuration key, got: %T", val)
}
f.apiToken = strVal
```

Also, it might be useful to trim spaces from strings such as the API token, URL, etc., using strings.TrimSpace to avoid potential configuration errors caused by accidental spaces.